openassessment/templates/openassessmentblock/oa_submission_answer.html

```diff
@@ -8,7 +8,7 @@ <h5 class="submission__answer__part__text__title">{% trans "The question for thi
             </div>
             <article class="submission__answer__part__prompt">
                 <div class="submission__answer__part__prompt__value">
-                    {{ part.prompt.description|linebreaks }}
+                    {% autoescape off %}{{ part.prompt.description }}{% endautoescape %}
```

Comment:
@sandroroux I agree that this part requires additional development. There are few ideas how to implement backward compatibility: 1. To add new option to the ORA settings what type of prompt should be used (simple text or html). By default it will be set as simple text. 2. The most easiest way - to check the first letter in the prompt - if it is `<` we may assert that this field is HTML. Otherwise use old construction `{{ part.prompt.description|linebreaks }}`. What do you think? May be you have more elegant solutions?

Reply:
Great ideas. I prefer option 1. Do you mind adding an ORA setting for HTML prompts?

Reply:
@sandroroux I suggest to add "HTML prompts" selector with options "Yes"/"No" ("No" is default) here (over the "Text Response" option):
![ora](https://user-images.githubusercontent.com/1876893/34231827-9ce4fc8c-e5ee-11e7-9072-12b3e9261997.jpeg)

Do you agree with such solution?

Reply:
Do we need the user to set it? Can't this be a hidden setting we set automatically if using the HTML editor?

Reply:
@sandroroux @dpowell  Another one solution (without changing studio UI):
1. We always use HTML editor for prompts in studio (i.e ORA settings).
2. We just change existing XBlock `prompt` to:
```
    prompt = String(
        default=None,
        scope=Scope.content,
        help="Do not use. For backwards compatibility only."
    )
``` 
3. And add new XBlock property like:
```
    prompt_html = String(
        default=DEFAULT_PROMPT,
        scope=Scope.content,
        help="The HTML prompts to display to a student."
    )
4. If `prompt` is not `None` it means that text prompt type is used - in this case we use `{{ part.prompt.description|linebreaks }}` to display description in the LMS + use encoding html tags and conversion new lines to `<br>` to display it in the ORA settings.
5. If prompt is `None` it means that new html prompt type is used - we use `{% autoescape off %}{{ part.prompt.description }}{% endautoescape %}` to display description in the LMS

Reply:
@dmitry-viskov, I'll speak with my colleagues about how to best approach this. Many of them are out of the office for the holidays. I'll  respond with a recommendation as soon as I can. Thank you very much for your ideas!

Reply:
@dmitry-viskov Is there anything that prevents us from formatting "part.prompt.description" before it's injected into the template?

Where "description" is defined, can we can change plaintext into HTML?

Reply:
@sandroroux yes, we can. But as for my opinion it's not important. it doesn't solve backward compatibility question.

Reply:
@dmitry-viskov Do you mind explaining why it wouldn't solve backwards compatibility? 

As I've understood it, removing `linebreaks` threatens backwards compatibility because plaintext `description`s will not be HTML formatted. Plaintext newlines must be converted into corresponding HTML tags before being rendered in the browser. Can we perform the conversion before the template expands `part.prompt.description`?

Could you point me to the code that generates "part.prompt.description"? The codepath may fork for HTML and plaintext cases. Can we detect between plaintext and HTML at `part.prompt.description`'s assignment?

Reply:
@sandroroux it doesn't matter where you make html encoding - in templates or in views or in other place. The problem is that some old prompts could contain HTML that should be displayed as HTML. line breaks is as particular case of such situation.

>Could you point me to the code that generates "part.prompt.description"? 

https://github.com/edx/edx-ora2/blob/master/openassessment/xblock/data_conversion.py#L104

Reply:
so the main problem is how to determinant which prompt is used - old (created for the plain text representation) or the new one (created for the HTML representation). I've suggested solution with the new XBlock property: https://github.com/edx/edx-ora2/pull/1058#discussion_r158165750 (see above)

Reply:
@dmitry-viskov I see. I thought tinyMCE wouldn't render newline characters as line breaks. Seeing as tinyMCE automatically formats plaintext line-breaks, let's tackle the backwards compatibility issue that you identified: unescaped HTML tags in plaintext prompts.

What happens when tinyMCE renders a prompt with unescaped HTML? How badly does ORA break? Can you post a screenshot of what happens?

In the meantime, I'll look into how common this is in edX courses.

Reply:
>Seeing as tinyMCE automatically formats plaintext line-breaks, let's tackle the backwards compatibility

I think you understood me wrong. tinyMCE doesn't automatically format plaintext line-breaks. But it is not a big problem because 1. when ORA block is published and there are some answers - xblock settings can't be changed, 2. when user will open settings - he will see that prompt is broken in the WYSIWYG editor. The main problem is in the prompt display.

Example 1:

![s1](https://user-images.githubusercontent.com/1876893/34718314-3b5c359e-f547-11e7-95a2-9674e2445ed5.jpeg)

![s2](https://user-images.githubusercontent.com/1876893/34718322-3fb089d8-f547-11e7-835b-4d24d53263b8.jpeg)

Example 2:

![s3](https://user-images.githubusercontent.com/1876893/34718331-46a8e528-f547-11e7-926e-975ce6cfc412.jpeg)

![s4](https://user-images.githubusercontent.com/1876893/34718335-4ab0c5b4-f547-11e7-9386-aa5c3faeed5a.jpeg)

Reply:
@sandroroux

I'm going to implement solution described in: https://github.com/edx/edx-ora2/pull/1058#discussion_r158165750

because as for my opinion this is the only way how to support backward compatibility for all cases

Reply:
@dmitry-viskov Thanks for illustrating the backwards compatibility concerns. I'm sorry for the misunderstanding. Please hold off on implementing a solution.

My colleagues and I are considering how to introduce your feature. We are assessing a few options. I'll let you know how to proceed shortly. 

In the meantime, thanks for being patient with me.